# MC316 Final Exam – Part B: PYTHON

* You will have 1 hour and 40 minutes to finish the test.
* This part includes 10 questions. Each question is worth 15 points.
* You get extra 30 points for full answers.
* Please read the instructions carefully. 

## I. SET UP


### Import pandas library for use in scripting. 

In [1]:
import pandas as pd

### Import the datasets and name them.
For this test, you will analyze revenue and expenditure data released by Kansas Department of Administration for the fiscal year 2018. While the original data can be directly downloaded [here](http://kanview.ks.gov/DataDownload.aspx), for your convenience, the instructor had them uploaded to a server at the following urls. 

#### For Kansas State University:
1. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv
2. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv

#### For University of Kansas:
3. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv
4. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv

You will need to import all 4 datasets and name them in an easy way to call out later.

*Please refer to the [Glossary](http://kanview.ks.gov/Glossary.aspx#f) for data description.

In [34]:
KSUEX = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv")

In [35]:
KSUREV = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv")

In [36]:
KUEX = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv")

In [37]:
KUREV = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv")

## II. REQUIRED QUESTIONS
* Make sure you type the answers in the cells right below the codes.
* If you need to add extra cells, click the + button in the toolbar up the top. 

### Q1: What is the total revenue recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total revenue recorded for University of Kansas in the same year?

In [39]:
KSUREV.Total.sum()

854324643.84

In [40]:
KUREV.Total.sum()

808767902.55

In [41]:
854324643.84 - 808767902.55

45556741.29000008

Answer: The total revenue in the fiscal year of 2018 for Kansas State University was 854,324,643.84 dollars. It is 45,556,741.29 higher then the total revenue for the University of Kansas that year.

### Q2: What is the total expense recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total expense recorded for University of Kansas in the same year?

In [42]:
KSUEX.Total.sum()

1018584373.02

In [43]:
KUEX.Total.sum()

990238875.6099999

In [44]:
1018584373.02 - 990238875.6099999

28345497.410000086

Answer: The total expense recorded for Kansas State in 2018 was 1,018,584,373.02 dollars. The difference with the total expense for Kansas was 28,345,497.410000086 dollars.

### Q3: Where did the two universities' primary revenues come from? 
*Refer to the description [here](http://kanview.ks.gov/Revenues/AcctCodeRevenues.aspx) to write your answer.

In [65]:
KSUREV.Primary_Account_Code_Descr.value_counts()

AGENCY EARNINGS                   637
REV FROM USE OF MONEY & PROP      354
REVENUE TRANSFERS                 341
NON REVENUE RECEIPTS              259
OTHER REVENUE                     160
GIFT DONATION FEDERAL GRANTS      152
TAXES                               7
Name: Primary_Account_Code_Descr, dtype: int64

In [66]:
KUREV.Primary_Account_Code_Descr.value_counts()

AGENCY EARNINGS                   376
REV FROM USE OF MONEY & PROP      251
NON REVENUE RECEIPTS              157
REVENUE TRANSFERS                 153
OTHER REVENUE                      57
TAXES                              45
GIFT DONATION FEDERAL GRANTS       38
Name: Primary_Account_Code_Descr, dtype: int64

In [67]:
637 - 376

261

Answer: The primary revenue for Kansas State was Agency Earnings at 637, and that is 261 more then Kansas. 

### Q4: What was the biggest source of revenues for Kansas State University? Compare to the one of University of Kansas.

Remove hashtag and run the following code to suppress scientific notation:

In [89]:
pd.set_option('display.float_format', lambda x: '%.3f' % x) 

In [90]:
KSUREV.groupby("Intermediate_Account_Code_Descr").amount.sum()

AttributeError: 'DataFrameGroupBy' object has no attribute 'amount'

In [91]:
merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 436413 entries, 0 to 436412
Data columns (total 43 columns):
Fiscal_Year_x                        436413 non-null int64
Fiscal_Quarter_x                     436413 non-null int64
Business_Unit_x                      436413 non-null int64
Business_Unit_Descr_x                436413 non-null object
GAAP_Fund_Type_x                     436413 non-null int64
GAAP_Fund_Type_Descr_x               436413 non-null object
Fund_Code_x                          436413 non-null int64
Fund_Code_Descr_x                    436413 non-null object
Budget_Ref_x                         436413 non-null int64
Budget_Ref_Descr_x                   436413 non-null object
Program_Code_x                       436413 non-null int64
Program_Code_Descr_x                 436413 non-null object
State_Function_x                     436413 non-null int64
State_Function_Descr_x               436413 non-null object
Primary_Account_Code_x               436413 non-null int6

In [94]:
KSUREV.groupby("Primary_Account_Code_Descr").Total.sum()


Primary_Account_Code_Descr
AGENCY EARNINGS                  412565112.820
GIFT DONATION FEDERAL GRANTS     278778848.030
NON REVENUE RECEIPTS              71947564.050
OTHER REVENUE                      8432030.160
REV FROM USE OF MONEY & PROP      34139318.040
REVENUE TRANSFERS                 48458616.490
TAXES                                 3154.250
Name: Total, dtype: float64

In [95]:
KUREV.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
AGENCY EARNINGS                  408409337.320
GIFT DONATION FEDERAL GRANTS     193836668.320
NON REVENUE RECEIPTS              69753247.670
OTHER REVENUE                     68241930.500
REV FROM USE OF MONEY & PROP      40950779.540
REVENUE TRANSFERS                 26013855.400
TAXES                              1562083.800
Name: Total, dtype: float64

Answer: The biggest source of revenue for Kansas State is agency earnings is 412,565,112.820, compared to Kansas' of 408,409,337.320.

### Q5: What accounts contributed to Kansas State University's biggest source of revenues?
If you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [98]:
KSUREV["Primary_Account_Code_Descr"] = KUREV["Primary_Account_Code_Descr"].str.strip()

After cleaning the spaces, you can start filtering your datasets to find the answer for Q5.

In [100]:
KSUREV["Primary_Account_Code_Descr"] = KUREV["Primary_Account_Code_Descr"].str.strip()

Answer: Agency Earnings, Gift Donation Federal Grants, Non Revenue Receipts, Other Revenue, REV from Use of Money & Prop, Revenue Transfers, and Taxes

### Q6: Where did the money go? (Answer for both universities)
*Refer to the description [here] (http://kanview.ks.gov/Expenditures/AcctCodeExpenditures.aspx) to write your answer.

In [119]:
KSUEX.groupby("Detail_Account_Code_Descr").Total.sum().reset_index().sort_values("Total", ascending=False)

Detail_Account_Code_Descr         Total
209  UNCLASSIFIED REGULAR           305142755.620
133  OTHER LOANS                    122486413.000
194  STUDENT SCHOLAR FELLOW GRANT    74721592.990
20   BUILDINGS                       39972418.640
68   GROUP HEALTH HOSPITALIZATION    36281335.170
50   EDUCATIONAL INST EE RETIRE SYS  21783012.730
58   FED AID TO QUALIFIED NON STATE  20045451.550
119  OASDI                           19126303.800
128  OTHER FEES                      18195826.220
210  UNCLASSIFIED TEMPORARY          18051482.160
192  STUDENT EMPLOYEES HOURLY        16409907.550
193  STUDENT RESEARCH ASSISTANTS     16146554.180
43   DEBT SVC TRANSFER PRINCIPAL     16011010.600
17   BOND PRINCIPAL                  15960000.000
16   BOND INTEREST                   15417585.700
42   DEBT SVC TRANSFER INTEREST      14662433.470
126  OTHER CONTRACTUAL SERVICES      14349200.250
51   ELECTRICITY                     12156496.410
159  PRO SCIENTIFIC SUPPLY OTHER     12095845.510
205  UNCLASS HOLIDAY INCL IN BASE    11109574.960
67   GRADUATE TEACHING ASSISTANTS    10441437.600
169  REPAIR AND SVC BLDG GROUND       9662271.180
136  OTHER PROFESSIONAL FEES          8317521.100
135  OTHER NON EXPENSE ITEMS          7789230.430
41   DATABASE ACCESS FEES NOT OITS    7135429.320
158  PRO AND SCIENTIFIC EQUIP         6935124.510
80   INTERFUND TRANSFER OUT           6271685.690
162  PUBLIC EE RETIREMENT SYSTEM      6242569.040
54   ENROLL FEES SPONS INST GRANT     5737654.200
63   FOOD FOR HUMAN CONSUMPTION       5628672.570
..                              ...           ...
44   DEPR LAND IMPROVEMENTS NON CAP      8680.690
200  TELECOM TRANSMISSION NON CAP        7195.570
85   INTERGOV LD COMM SVC OP OITS        6097.040
153  PRINTING EQUIP NON CAPITAL          5764.100
47   DRUGS AND PHARMACEUTICALS           5505.580
130  OTHER FUEL                          4068.040
161  PROPERTY DAMAGE OR LOSS CLAIMS      3704.280
64   FUEL OIL DIESEL AND KEROSENE        3556.830
175  ROYALTY PAYMENTS                    3028.080
29   COMPUTER EQUIPMENT RENTAL           2730.450
160  PROPANE BUTANE NATURAL GAS          2692.920
151  PRINT BIND SURCHARGE ST ONLY        2630.220
56   FCIP-PRINCIPAL-BI-C/I               1884.780
12   ASSESSMENTS FOR IMPROVEMENTS        1778.080
53   EMPLOYEE AWARDS NOT IRS REPORT      1764.370
110  MOVING EE PERS EFFECTS IN ST         772.650
101  LOCAL FUNDS SHORTAGES                441.890
188  STATE CAR EXPENSE-INTERNATL          414.400
14   BANK FEES                            292.700
83   INTERGOV LD COMM SVC CAP OITS        152.410
30   COMPUTER PROGRAMMING CAP OITS          3.720
24   CLASSIFIED REGULAR                     0.000
173  RETURN OF UNEXPENDED FED GRANT         0.000
134  OTHER MISC ADVANCES                    0.000
124  OTHER CLAIMS                           0.000
97   LIVESTOCK                          -1195.540
52   EMPLOYEE AWARDS IRS REPORTABLE     -6278.370
82   INTERGOV LCL COMM SVC OP OITS      -8649.140
113  NATIONAL DIRECT STUDENT LOANS     -98046.440
149  PRINCIPAL AND PREMIUMS           -175082.020

[217 rows x 2 columns]

In [122]:
KUEX.groupby("Detail_Account_Code_Descr").Total.sum().reset_index().sort_values("Total", ascending=False)

Detail_Account_Code_Descr         Total
237  UNCLASSIFIED REGULAR           305121246.380
153  OTHER LOANS                    155585941.820
223  STUDENT SCHOLAR FELLOW GRANT    71155246.600
26   BUILDINGS                       41133478.280
79   GROUP HEALTH HOSPITALIZATION    35055117.340
147  OTHER FEES                      26918847.290
143  OTHER CONTRACTUAL SERVICES      25602259.510
64   EDUCATIONAL INST EE RETIRE SYS  23518974.990
34   CLASSIFIED REGULAR              18906657.600
78   GRADUATE TEACHING ASSISTANTS    18772266.860
134  OASDI                           18479835.850
19   BOND INTEREST                   13514706.920
65   ELECTRICITY                     12609894.380
222  STUDENT RESEARCH ASSISTANTS     12100134.710
20   BOND PRINCIPAL                  11795000.000
58   DEBT SVC TRANSFER PRINCIPAL     11605000.000
221  STUDENT EMPLOYEES HOURLY        10739204.610
66   EMPLOYEE AND STUDENT TUITION    10055763.700
57   DEBT SVC TRANSFER INTEREST       9736916.160
62   DUES AND SUBSCRIPTIONS           7985473.750
194  REPAIR & SVC OTHER               7462877.120
23   BUILDING IMPROVEMENTS            6738896.190
234  UNCLASS HOLIDAY INCL IN BASE     6289556.090
18   BLDGS AND GROUNDS MATERIALS      4865591.890
120  MEDICARE                         4636221.930
156  OTHER PROFESSIONAL FEES          4424959.140
188  PUBLIC EE RETIREMENT SYSTEM      4256270.700
185  PRO SCIENTIFIC SUPPLY OTHER      4164762.930
12   ARCHITECTS AND ENGINEERS         4004007.880
238  UNCLASSIFIED TEMPORARY           3878461.940
..                              ...           ...
104  INTERGOV POSTAGE OP OITS             991.570
102  INTERGOV OTHER COMM NOT OITS         941.880
187  PROPERTY DAMAGE OR LOSS CLAIMS       823.710
215  ST BLDGS OPERATING CHARGE            769.000
73   FEED AND FORAGE                      637.160
154  OTHER MISC ADVANCES                  630.000
220  STEAM                                593.230
86   HOSPITALS AND NURSING HOMES          500.000
89   HOUSEHOLD FURNITURE                  447.560
186  PROPANE BUTANE NATURAL GAS           334.400
202  SELF PROP SVC OIL CHANGE             290.410
108  LAB RANDOM EE DRUG TEST              167.800
69   ER KPERS WAR 3RD PTY/CNSLNT          118.140
16   ATTORNEYS AND LAWYERS                117.330
43   COMPUTER PROGRAMMING CAP OITS        114.690
99   INTERGOV LD COMM SVC CAP OITS         95.960
10   ANIMAL DRUGS                          83.470
149  OTHER FUEL                            66.920
124  MOVING EE PERS EFFECTS IN ST          60.630
217  STATE CAR EXPENSE-INTERNATL           52.500
92   HYDRAULIC OIL                         48.410
116  LOAN INTEREST                         35.110
205  SELF PROP TIRE REPAIR INSTALL         14.000
107  LAB RANDOM ALCOHOL TEST                9.000
13   ASSESSMENTS FOR IMPROVEMENTS           0.000
232  TRUCKS NON CAP                         0.000
67   ENROLL FEES SPONS INST GRANT           0.000
2    ADVERTISING AND MARKETING              0.000
71   FCIP-PRINCIPAL-BI-C/I             -14209.000
98   INTERGOV LCL COMM SVC OP OITS   -3496362.680

[242 rows x 2 columns]

Answer: For both Universities, the main two places where the money went was Unclassified Regular and Other Loans. 

### Q7: What costs two universities the most? 

In [145]:
KUEX.groupby("Detail_Account_Code_Descr").Total.sum().reset_index().sort_values("Total", ascending=False)

Detail_Account_Code_Descr         Total
237  UNCLASSIFIED REGULAR           305121246.380
153  OTHER LOANS                    155585941.820
223  STUDENT SCHOLAR FELLOW GRANT    71155246.600
26   BUILDINGS                       41133478.280
79   GROUP HEALTH HOSPITALIZATION    35055117.340
147  OTHER FEES                      26918847.290
143  OTHER CONTRACTUAL SERVICES      25602259.510
64   EDUCATIONAL INST EE RETIRE SYS  23518974.990
34   CLASSIFIED REGULAR              18906657.600
78   GRADUATE TEACHING ASSISTANTS    18772266.860
134  OASDI                           18479835.850
19   BOND INTEREST                   13514706.920
65   ELECTRICITY                     12609894.380
222  STUDENT RESEARCH ASSISTANTS     12100134.710
20   BOND PRINCIPAL                  11795000.000
58   DEBT SVC TRANSFER PRINCIPAL     11605000.000
221  STUDENT EMPLOYEES HOURLY        10739204.610
66   EMPLOYEE AND STUDENT TUITION    10055763.700
57   DEBT SVC TRANSFER INTEREST       9736916.160
62   DUES AND SUBSCRIPTIONS           7985473.750
194  REPAIR & SVC OTHER               7462877.120
23   BUILDING IMPROVEMENTS            6738896.190
234  UNCLASS HOLIDAY INCL IN BASE     6289556.090
18   BLDGS AND GROUNDS MATERIALS      4865591.890
120  MEDICARE                         4636221.930
156  OTHER PROFESSIONAL FEES          4424959.140
188  PUBLIC EE RETIREMENT SYSTEM      4256270.700
185  PRO SCIENTIFIC SUPPLY OTHER      4164762.930
12   ARCHITECTS AND ENGINEERS         4004007.880
238  UNCLASSIFIED TEMPORARY           3878461.940
..                              ...           ...
104  INTERGOV POSTAGE OP OITS             991.570
102  INTERGOV OTHER COMM NOT OITS         941.880
187  PROPERTY DAMAGE OR LOSS CLAIMS       823.710
215  ST BLDGS OPERATING CHARGE            769.000
73   FEED AND FORAGE                      637.160
154  OTHER MISC ADVANCES                  630.000
220  STEAM                                593.230
86   HOSPITALS AND NURSING HOMES          500.000
89   HOUSEHOLD FURNITURE                  447.560
186  PROPANE BUTANE NATURAL GAS           334.400
202  SELF PROP SVC OIL CHANGE             290.410
108  LAB RANDOM EE DRUG TEST              167.800
69   ER KPERS WAR 3RD PTY/CNSLNT          118.140
16   ATTORNEYS AND LAWYERS                117.330
43   COMPUTER PROGRAMMING CAP OITS        114.690
99   INTERGOV LD COMM SVC CAP OITS         95.960
10   ANIMAL DRUGS                          83.470
149  OTHER FUEL                            66.920
124  MOVING EE PERS EFFECTS IN ST          60.630
217  STATE CAR EXPENSE-INTERNATL           52.500
92   HYDRAULIC OIL                         48.410
116  LOAN INTEREST                         35.110
205  SELF PROP TIRE REPAIR INSTALL         14.000
107  LAB RANDOM ALCOHOL TEST                9.000
13   ASSESSMENTS FOR IMPROVEMENTS           0.000
232  TRUCKS NON CAP                         0.000
67   ENROLL FEES SPONS INST GRANT           0.000
2    ADVERTISING AND MARKETING              0.000
71   FCIP-PRINCIPAL-BI-C/I             -14209.000
98   INTERGOV LCL COMM SVC OP OITS   -3496362.680

[242 rows x 2 columns]

Answer: Both Universities' "unclassified reglar" and "other loans" cost them the most. 

### Q8: Calculate the percentage of this cost over each university's total expense.

In [148]:
305142755.620/1018584373.02

0.299575335831319

In [150]:
305121246.380/990238875.6099999

0.3081289312056562

In [149]:
0.299575335831319 *100

29.9575335831319

K-State: 29.9%
KU: 30.8%

### Q9: Breakdown the cost you found in Q7 for both universities.
Again, if you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [3]:
#yourdataframe["yourcolumnname"] = yourdataframe["yourcolumnname"].str.strip()

<b>Answer:</b>

### Q10: Compare the breakdown costs from Q9 side by side.
When you merge the two subsets, make sure to add the attribute <b>how ="outer"</b> so that you can keep every item from both sets.
*For example: merge = pd.merge(df1, df2, how="outer", on="columnname")

<b>Answer:</b>

## III. BONUS QUESTIONS
* Each question's worth 15 points.


### B1: Which university made a profit in the fiscal year of 2018?

In [68]:
KSUREV.Total.sum()

854324643.84

In [69]:
KSUEX.Total.sum()

1018584373.02

In [70]:
KUREV.Total.sum()

808767902.55

In [71]:
KUEX.Total.sum()

990238875.6099999

In [72]:
854324643.84 - 1018584373.02

-164259729.17999995

In [73]:
808767902.55 - 990238875.6099999

-181470973.05999994

Answer: Neither University made a profit in 2018.

### B2: Which software program do you like the most in this course? Why?

Answer: I like tableau the best, because of the cool graphics that we were able to make through it. The graphs are more interesting to look at and interact with then just normal graphs. It was also easier for me to use then some of the other software programs we used in class, making it more enjoyable in that sense.

### END OF THE TEST